src/domain/accounts/services/src/access_token_service_impl.rs

```diff
@@ -0,0 +1,182 @@
+#[scope(Singleton)]
+#[interface(dyn AccessTokenService)]
+impl AccessTokenServiceImpl {
+    #[allow(clippy::needless_pass_by_value)]
+    pub fn new(access_token_repository: Arc<dyn AccessTokenRepository>) -> Self {
+        Self {
+            access_token_repository,
+        }
+    }
+
+    pub fn generate_access_token_impl(&self) -> KamuAccessToken {
+        let mut random_token_bytes = [0_u8; ACCESS_TOKEN_BYTES_LENGTH];
+        rand::thread_rng().fill(&mut random_token_bytes);
+
+        let token_checksum = crc32fast::hash(&random_token_bytes);
+        let token_id = Uuid::new_v4();
+        let full_token_bytes = [
+            token_id.as_bytes(),
+            random_token_bytes.as_slice(),
+            token_checksum.to_be_bytes().as_slice(),
+        ]
+        .concat();
+        let base32_token = base32::encode(base32::Alphabet::Crockford, &full_token_bytes);
+
+        KamuAccessToken {
+            prefix: ACCESS_TOKEN_PREFIX.to_string(),
+            id: token_id,
+            random_bytes: random_token_bytes,
+            random_bytes_hash: Multihash::from_digest_sha3_256(&random_token_bytes),
+            checksum: token_checksum,
+            composed_token: format!("{ACCESS_TOKEN_PREFIX}_{}", &base32_token),
+            base32_token,
+        }
+    }
+
+    pub fn decode_token_impl(
+        &self,
+        access_token: String,
+    ) -> Result<KamuAccessToken, EncodeTokenError> {
+        if access_token.len() != ENCODED_ACCESS_TOKEN_LENGTH {
+            return Err(EncodeTokenError::InvalidTokenLength);
+        }
+        if let Some((token_prefix, encoded_token)) = access_token.split_once('_') {
+            if token_prefix != ACCESS_TOKEN_PREFIX {
+                return Err(EncodeTokenError::InvalidTokenPrefix);
+            }
+
+            if let Some(decoded_token_bytes) =
+                base32::decode(base32::Alphabet::Crockford, encoded_token)
+            {
+                let token_id = Uuid::from_bytes(
+                    decoded_token_bytes[0..16]
+                        .try_into()
+                        .expect("Invalid uuid length"),
+                );
+                let token_body: [u8; 16] = decoded_token_bytes[16..32]
+                    .try_into()
+                    .expect("Invalid token_bytes_length");
+                let token_checksum: [u8; 4] = decoded_token_bytes[32..36]
+                    .try_into()
+                    .expect("Invalid checksum length");
+                let calculated_checksum = crc32fast::hash(&token_body);
+                if calculated_checksum.to_be_bytes() != token_checksum {
+                    return Err(EncodeTokenError::InvalidTokenChecksum);
+                }
+
+                return Ok(KamuAccessToken {
+                    id: token_id,
+                    prefix: token_prefix.to_string(),
+                    random_bytes: token_body,
+                    random_bytes_hash: Multihash::from_digest_sha3_256(&token_body),
```

Comment:
I think that this should not be recomputed. We only want the hash at the creation point of time. Maybe different token structs are needed for the create and decode cases.

Reply:
Should we even return this struct? Isn't ID enough to determine the account?